### PR TITLE
index: Fast path for EmptyPostings cases in Merge, Intersect and Without

### PR DESCRIPTION
**Pull request by bwplotka**
Updates old https://github.com/prometheus/tsdb/pull/233 onto fresh TSDB code.

closes https://github.com/prometheus/tsdb/pull/233

### CODE REVIEW

**Comment by bwplotka** [comment]
ping @gouthamve

**Comment by krasi-georgiev** [comment]
@bwplotka 	I did a PR against your branch with few adjustments.
https://github.com/bwplotka/tsdb/pull/2

**Comment by krasi-georgiev** [comment]
LGTM, Thanks!

**Comment by krasi-georgiev** [comment]
@codesome 	want to have a quick look before I merge it?

**Comment by krasi-georgiev** [comment]
ping @bwplotka

**Comment by bwplotka** [comment]
Updated @krasi-georgiev and @codesome 

Thanks for ping & reviews. Ping me directly on IRC or slack if you want my review/PR update I am flooded with GH notifications ):

I changed a lot now, because of conflicts - I rewrote test cases to make more sense, let me know if that is OK with you guys.

File: index/postings.go [hunk 1]
```diff
@@ -395,18 +409,24 @@ type mergedPostings struct {
...
 			if it.Err() != nil {
-				return &mergedPostings{err: it.Err()}
+				return &mergedPostings{err: it.Err()}, true
 			}
 		}
 	}
-	return &mergedPostings{h: ph}
+
+	if len(ph) == 0 {
+		return nil, false
```
Comment (krasi-georgiev) [inline index/postings.go]: shouldn't this be before anything else?

File: index/postings_test.go [hunk 2]
```diff
@@ -61,39 +61,125 @@ func TestMemPostings_ensureOrder(t *testing.T) {
...
+				toPostings(),
+			},
+			res: toPostings(),
 		},
 	}
 
 	for _, c := range cases {
-		a := newListPostings(c.a)
-		b := newListPostings(c.b)
+		t.Run("", func(t *testing.T) {
```
Comment (krasi-georgiev) [inline index/postings_test.go]: maybe add some indication to know when a test fails which case failed.
```
map[string]struct {
.
.

for t, c := range cases {
		t.Run(t, func(t *testing.T) {
```

File: index/postings_test.go [hunk 3]
```diff
@@ -61,39 +61,125 @@ func TestMemPostings_ensureOrder(t *testing.T) {
...
+			i := Intersect(c.in...)
+
+			if c.res == EmptyPostings() {
+				testutil.Equals(t, EmptyPostings(), i)
+				return
+			}
+
+			if i == EmptyPostings() {
+				t.Fatal("intersect unexpected result: EmptyPostings sentinel")
+			}
```
Comment (krasi-georgiev) [inline index/postings_test.go]: why do we need?
```
if c.res == EmptyPostings() {
	testutil.Equals(t, EmptyPostings(), i)
	return
}

if i == EmptyPostings() {
	t.Fatal("intersect unexpected result: EmptyPostings sentinel")
}
```

the test passes even without these extra checks.

File: index/postings_test.go [hunk 4]
```diff
@@ -541,3 +698,81 @@ func TestIntersectWithMerge(t *testing.T) {
...
+			base: toPostings(1, 2, 3),
+			drop: toPostings(3, 4, 5),
+
+			res: toPostings(1, 2),
+		},
+	}
+
+	for _, c := range cases {
+		t.Run("", func(t *testing.T) {
+			if c.res == nil {
```
Comment (krasi-georgiev) [inline index/postings_test.go]: don't think we need this. A panic is ok in a test in my opinion

File: index/postings.go (hunk 1 above)
Comment (bwplotka) [reply index/postings.go]: hm? We know this only after `it.Next` bit.

File: index/postings_test.go (hunk 2 above)
Comment (bwplotka) [reply index/postings_test.go]: not needed. Passing `""` cause go test to put a number, so exactly what you propose.

File: index/postings_test.go (hunk 3 above)
Comment (bwplotka) [reply index/postings_test.go]: Exactly to expect `EmptyPostings` explicitly!

File: index/postings_test.go (hunk 4 above)
Comment (bwplotka) [reply index/postings_test.go]: If you think so, can remove

File: index/postings_test.go (hunk 2 above)
Comment (krasi-georgiev) [reply index/postings_test.go]: yes that is right but when you have many cases it is kind of difficult to count which one failed exactly.

File: index/postings.go (hunk 1 above)
Comment (krasi-georgiev) [reply index/postings.go]: yep , I missed that bit.

File: index/postings_test.go [hunk 5]
```diff
@@ -197,35 +283,106 @@ func TestMultiMerge(t *testing.T) {
...
+				EmptyPostings(),
+				EmptyPostings(),
+				EmptyPostings(),
+				EmptyPostings(),
+			},
+			res: EmptyPostings(),
+		},
+		{
+			in: []Postings{
+				toPostings(1, 2, 3, 4, 5),
```
Comment (krasi-georgiev) [inline index/postings_test.go]: what is blocking to use `newListPostings` everywhere?

File: index/postings_test.go (hunk 3 above)
Comment (krasi-georgiev) [reply index/postings_test.go]: yep, now I see it.

File: index/postings_test.go (hunk 5 above)
Comment (bwplotka) [reply index/postings_test.go]: boilerplate

File: index/postings.go [hunk 6]
```diff
@@ -382,13 +388,21 @@ func Merge(its ...Postings) Postings {
...
+
 	sort.Slice(pl, func(i, j int) bool { return pl[i] < pl[j] })
 	return newListPostings(pl)
 }
 
 // Without returns a new postings list that contains all elements from the full list that
-// are not in the drop list
+// are not in the drop list.
 func Without(full, drop Postings) Postings {
+	if full == EmptyPostings() {
```
[the remark this answers is not in the document]
Comment (krasi-georgiev) [reply index/postings.go]: nope, drop are the ones to be removed

File: index/postings_test.go (hunk 2 above)
Comment (bwplotka) [reply index/postings_test.go]: Disagree, let's not obfuscate the tests more -> everyone can count

Comment (krasi-georgiev) [reply index/postings_test.go]: that is why I added the title in:
https://github.com/prometheus/tsdb/pull/559/files

Comment (bwplotka) [reply index/postings_test.go]: Yea, I agree there is a value for this, but I am afraid I won't have time to do this, this week ): 

I would suggesting merging this PR with just indexing (numbers), then we can work on making it more verbose later on. What do you think?

File: index/postings_test.go (hunk 5 above)
Comment (krasi-georgiev) [reply index/postings_test.go]: how about if you change 
`newListPostings([]uint64) -> newListPostings(list ...uint64)`
than you can call it as `newListPostings(1, 2, 3, 4, 5)`
just tried it and it removes that boilerplate without the need for an extra func.


not a blocker though can leave it as if you prefer.

File: index/postings_test.go (hunk 2 above)
Comment (krasi-georgiev) [reply index/postings_test.go]: sure I don't mind can always add it later if it gets annoying to count :smile:

File: index/postings_test.go (hunk 5 above)
Comment (bwplotka) [reply index/postings_test.go]: It will increase this PR size, but did it, I like this idea :+1:

**Comment by krasi-georgiev** [comment]
Thanks Bartek!

**Comment by bwplotka** [comment]
Thanks Krasi (:

**Pull request merged**